### ISSUE TITLE
Some notes on peersets and current problems

### ISSUE THREAD

**Issue by hsanjuan** [ipfs/ipfs-cluster#219]
These are some notes about some issues which I need to put aside for a bit, so I don't forget.

On the wake of #131, I have revisited much of the code dealing with:

* PeerAdd
* PeerRm
* Join/bootstrap
* Leave

There was and there is a problem right now with how these operations are approached.

Right now, ipfs-cluster's `peerManager` keeps track of a map of `pid->multiaddress`. The current cluster peers are the PIDs in that map. This map determines:

* The current set of peers when performing broadcast operations
* The multiaddresses of these peers when adding a new peer to cluster (we push this list to the newcomer)

However, `raft` also keeps it's own peerset as part of the `Configuration` object. This peeerset, due to its nature (`ServerID/ServerAddress`) fields, only contains peer IDs.

And of course, the libp2p host also keeps an address book with peer IDs and addresses.

As a result, we are tracking the peer-set in 3 different places in some way or another. This means that adding a new member to cluster needs to:

* Add the Peer ID and Multiaddress to our peerManager
* Provide Raft with a new peer ID and commit that operation
* Make sure the libp2p host peerstore has it too.

Here are some reasons why this was done so (not all them need to be good reasons):

* We cannot fully rely on Raft peerstore, because it doesn't know the multiaddresses, so when I new peer joins, we need to send them.
* We cannot fully rely on libp2p peerstore, because it tends to add and keep entries for nodes (i.e. just one node contacting another will add an entry). It is very difficult to be sure of the right peerset only from the libp2p peerstore.
* The peer manager is therefore a peerstore but only for our peers.
* We need to know when a peer leaves, so we can shut it down among other things. There is no easy way to observe  changes to the Raft `Configuration`
* Even if we could observe membership changes in Raft (we can by polling constantly), the Consensus component is separate and peer membership was something that we wanted to manage in the main Cluster component.
* Observing the Raft peerset requires polling raft for it which is [probably] considerably more expensive than just checking our own copy (maybe this is not a true assumption).

Currently, peerManager's peerset updates are done via consensus log entries (which are not saved to any state). This is very ugly because Raft already creates a log entry for the `Configuration` peerset. And we just do another one on top for our own peerset. This was done so (and not with a simple broadcast) because this way we could make sure that an addition was seen by all peers (and even picked up later if a peer was down).

**There are multiple problems with the current ways**:

* A bunch of tricky code
* Hard coordination between raft's peerset and our own when adding and removing a peer
* Because the log entries for PeerAdd and PeerRm are ephemeral (we don't save them to the state), a node that has been down long enough time for raft to make an intermediary snapshot will sure miss LogAdd/LogRm entries (as they will not be replayed). Thus it will enter a very gray inconsistent area.
* We have no idea when we are in an unconsistent peerset state, where the peerManager has a set of peers and the consensus another one.
* When re-joining an already removed peer (with PeerAdd), the log entry to remove that peer is replayed to the joining member. This results in that peer removing itself.
* PeerAdd operation where a standalone-node joins a cluster makes no sense anymore. We should always bootstrap, or at least never let a standalone node which will be added to a cluster elect a leader.
* When a peer joins, we need to commit stuff to Raft, which may require forwarding the operations to the master, and after all that works, and the peer is part of Raft, we need to inform the new peer of the multiaddresses of the rest of the people. This is known to fail when multiple peers are calling Join() on multiple cluster members. This results in some peers not learning about other peers which are joining at the exact same time.
* When a peer is removed, we have to be very careful to shut it down only after Raft has got the memo. If that peer happens to be the leader, and we shut it down immediately, other peers may not know about it (this works now, but with some acrobatics).
* The double-committing to raft log requires special code, special log entry handling...
* Currently multiple multiaddresses for the same peer are not supported

**Ideas to fix his**

Ideally the peerManager map should not exist. Raft `Configuration` object should be the source of truth giving a slice of current cluster peers. The libp2p host peerset will store all addresses for them (and any others). This allows:

* We have a source of truth for current peers via a Peers() method in the consensus component.
* We have their multiaddresses and can forward them when PeerAdd() happens. We may also have many other multiaddresses from old peers or whatnot. We just ignore those.

We however, lose "notification" on peer add/remove. I *think* (not sure), that this should be ok except for the case when a peer leaves and needs to be shutdown, for which we will need a explicit notification. A cluster peers needs to realize somehow if it's not part of the consensus anymore. If the shutdown notification comes from someone else, it may not arrive (what happens if the peer was down, is removed, then it's started again ?). Should we poll the configuration object to see if we are gone from it? What happens when we poll without a leader or bad network conditions? This needs a bit of figuring out what raft does.

This brings into question whether we need PeerAdd and PeerRm operations at all, or should we be happy with Join/Leave operations only. Join/Leave are easier to think of because they are initiated by the affected peer. PeerAdd() and PeerRm() are different because they are initiated by an unrelated peer and affect others. In the case of consul (main usecase of hashicorp raft), they content with equivalents for join/peerRm(forced-leave)/Leave. Forced-leave is a fallback option to remove a peer that is gone and hasn't left the cluster orderly.

Hopefully with this notes my future-self or someone can fix this for good.

**Comment by ZenGround0** [comment]
FWIW I am in favor of moving towards a Join/Leave/(forced)PeerRm() paradigm.

Something else that jumped out at me: the above mentions that one potential downside of removing the LogRmPeer and LogAddPeer ops is that cluster will no longer be able to get notified of additions and removals by consensus.  However this seems like a waste as the consensus goes through the trouble of including the Configuration peerset in the log.  I wonder if it would be possible to surface internal log ops (like modifying the Configuration peerset) to the user provided FSM (maybe through a separate `FSM.ApplyInternalTo` or something?  That way we could keep a single source of truth and still get notified.  I'm guessing this would need changes from hashicorp-raft so I am not suggesting it as an immediate fix, but it seems like behavior others might find useful too.

**Comment by hsanjuan** [comment]
Yeah, Raft has the notion of Observers and Observers should probably be able to observe for Configuration peerset updates and notify about them, but this is not available right now. The PR simply polls the peerset regularly to detect if we stopped being in it.
